File: src/main/java/net/dontdrinkandroot/gitki/service/markdown/DefaultMarkdownService.java
package net.dontdrinkandroot.gitki.service.markdown;

import org.apache.commons.lang3.StringUtils;
import org.commonmark.Extension;
import org.commonmark.ext.gfm.tables.TablesExtension;
import org.commonmark.node.Node;
import org.commonmark.parser.Parser;
import org.commonmark.renderer.html.HtmlRenderer;

import java.util.Arrays;
import java.util.List;

/**
 * @author Philip Washington Sorst <philip@sorst.net>
 */
public class DefaultMarkdownService implements MarkdownService
{
    private final Parser parser;
    private final HtmlRenderer renderer;

    public DefaultMarkdownService()
    {
        List<Extension> extensions = Arrays.asList(TablesExtension.create());
        this.parser = Parser.builder().extensions(extensions).build();
        this.renderer = HtmlRenderer.builder()
                .extensions(extensions)
                .attributeProviderFactory(new GitkiAttributeProviderFactory())
                .build();
    }

    @Override
    public String parseToHtml(String source)
    {
        if (StringUtils.isEmpty(source)) {
            return "";
        }

        return this.renderer.render(this.parse(source));
    }

    public Node parse(String source)
    {
        return this.parser.parse(source);
    }
}

File: src/main/kotlin/net/dontdrinkandroot/gitki/service/markdown/DefaultMarkdownService.kt
package net.dontdrinkandroot.gitki.service.markdown

import org.apache.commons.lang3.StringUtils
import org.commonmark.ext.gfm.strikethrough.StrikethroughExtension
import org.commonmark.ext.gfm.tables.TablesExtension
import org.commonmark.node.Node
import org.commonmark.parser.Parser
import org.commonmark.renderer.html.HtmlRenderer
import org.springframework.stereotype.Service
import java.util.*

@Service
class DefaultMarkdownService : MarkdownService {

    private val parser: Parser

    private val renderer: HtmlRenderer

    init {
        val extensions = Arrays.asList(TablesExtension.create(), StrikethroughExtension.create())
        parser = Parser.builder().extensions(extensions).build()
        renderer = HtmlRenderer.builder()
            .extensions(extensions)
            .attributeProviderFactory(GitkiAttributeProviderFactory())
            .build()
    }

    override fun parseToHtml(source: String): String =
        if (StringUtils.isEmpty(source)) "" else renderer.render(parse(source))

    fun parse(source: String?): Node = parser.parse(source)
}